### PR TITLE
[breaking] replace routeId with route.id

### LINKED ISSUE #6980: `routeId` not tracked as dependency to layout `load`
### Describe the bug

When you access `event.routeId` in a layout `load` function, the `load` function does not re-run when `routeId` changes. Since routeId is based on the URL, we should track dependencies to it similar to [url and params](https://kit.svelte.dev/docs/load#invalidation).

This probably only makes sense in a `layout` load. On `page` loads, routeId should stay the same.

### Reproduction

[StackBlitz](https://stackblitz.com/edit/sveltejs-kit-template-default-pfr37v?file=src/routes/+layout.server.js)

1. Load page. Notice routeId for the current page is displayed
2. Navigate to another page. The routeId does not change until a page reload

In this case you could work around the issue by retrieving it from the `$page` store, but if you're doing something else with `routeId`, that would not be re-run.

```js
// +layout.server.js
export const load = ({routeId}) => {
  return {
    routeId
  }
}
```

### Logs

```shell
n/a
```

### System Info

```shell
System:
    OS: Linux 5.0 undefined
    CPU: (4) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 0 Bytes / 0 Bytes
    Shell: 1.0 - /bin/jsh
  Binaries:
    Node: 16.14.2 - /usr/local/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 7.17.0 - /usr/local/bin/npm
  npmPackages:
    @sveltejs/adapter-auto: next => 1.0.0-next.80 
    @sveltejs/kit: next => 1.0.0-next.499 
    svelte: ^3.46.0 => 3.50.1 
    vite: ^3.1.0 => 3.1.3
```

### Severity

serious, but I can work around it

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by dummdidumm**
Closes #6980

Draft, because the tracking part is missing, and I'm not sure yet if we want to replace routeId in _all_ places (like `$page.routeId`)

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. All changesets should be `patch` until SvelteKit 1.0

### CODE REVIEW

**Comment by Conduitry** [comment]
Is there anything else we foresee living in `route`? `params` could get moved in there maybe?

Whatever we do, for consistency I'd be in favor of having `$page` be the same shape to the extent possible.

**Comment by dummdidumm** [comment]
`params` could be moved into it, but that would be too big of a breaking change at this point, and it also results in worse DX, because you'd have to destructure the input twice then.

**Comment by Rich-Harris** [comment]
The site is failing to deploy because a code snippet in the docs references `route.id` instead of `route?.id`. It'd be easy to fix, but I wonder if instead the types should reflect the fact that `route` always exists in the context of `load` or a `+server.js` handler?

**Comment by Rich-Harris** [comment]
I made that change for `LoadEvent` and `ServerLoadEvent`, trying to figure out the best approach for `RequestHandler`. If I make this change...

```diff
export interface RequestHandler<
	Params extends Partial<Record<string, string>> = Partial<Record<string, string>>
> {
-	(event: RequestEvent<Params>): MaybePromise<Response>;
+	(event: RequestEvent<Params> & { route: { id: string } }): MaybePromise<Response>;
}
```

...then it works, but this is what I get if I hover over `route`:

image

Not totally ideal. Do we need to come up with a new type that extends `RequestEvent`? Or should the route ID be a generic argument? (That way we could make it an actual value for the generated types, or a list of possible values for layouts, etc)

**Comment by dummdidumm** [comment]
I thought this, too, but in some cases the root layout is loaded but without a route - for example when an infinite redirect loop is detected, or in case of a 404 - many _ugh_-vibes on this PR 😅

**Comment by Rich-Harris** [comment]
Ah crap, you're right — will revert the last commit.

What if we went the generic route though, and made `route` always exist but with `id` defaulting to `string | null`?

```ts
// .svelte-kit/types/src/routes/blog/$types.d.ts
export type PageLoad<...> = Kit.Load<..., OutputData, 'blog'>;
export type LayoutLoad<...> = Kit.Load<..., OutputData, 'blog' | 'blog/[slug]'>;
```

Or is it misleading for the route object to exist when no route matched?

**Comment by dummdidumm** [comment]
Funny how we go through the same thought process independently. I asked myself the same question, and came to the conclusion that it's weird. Then again, all of this is weird, so we need to choose something that is the least weird while being easy enough to use. After having implemented this, part of me thinks `getRouteId()` would be better after all - unless you imagine any other things ending up in that object later on.
The generic is interesting, although I'm wondering if that typing maintenance is worth it.
What are people using `routeId` for anyway? Surely it's only in layouts?

**Comment by Rich-Harris** [comment]
And just as I'd come around to `route.id`! It's definitely more future-proof, but I can't imagine what it would need to be future-proof against (i.e. I can't think of any properties we'd need beyond `id`). Then again that's sort of the point of being future-proof — you can't anticipate everything.

I think parity with `$page` is a reasonably convincing argument. We _could_ have `event.getRouteId()` and `$page.routeId` but it would definitely feel a bit more natural to have `event.route.id` and `$page.route.id`.

> What are people using routeId for anyway? Surely it's only in layouts?

Yeah, I think the main use case is where you want some special case logic in a layout ('require the user to be logged in, unless the current `routeId` is `/login`', or whatever).

**Comment by dummdidumm** [comment]
I like `route.id` more for future-proofness and for how it looks when reading it. I'm not sure how I feel about `route: { id: string | null }` vs `route: { id: string } | null`. The latter feels more correct, but it's easier to produce runtime nullpointers with it. So ... yeah, maybe go with `route: { id: string | null }`

File: packages/kit/types/index.d.ts [hunk 1]
```diff
@@ -627,15 +635,17 @@ export interface ServerLoadEvent<
 
 export interface Action<
 	Params extends Partial<Record<string, string>> = Partial<Record<string, string>>,
-	OutputData extends Record<string, any> | void = Record<string, any> | void
+	OutputData extends Record<string, any> | void = Record<string, any> | void,
+	RouteId extends string | null = string | null
```
Comment (dummdidumm) [inline packages/kit/types/index.d.ts:639]: do we need the overhead of that interface on the action interfaces? Actions can only be used on pages, so routeId can only be one thing, so it's useless to use it, so probably noone does that, so we probably can just type this as `string` on `RequestEvent`

**Comment by Rich-Harris** [comment]
This autocompletion is kinda neat:

https://user-images.githubusercontent.com/1162160/199323074-7657d332-5e74-40fb-8681-8597f35ee530.mov

Also, this typechecking 😍 

image

File: packages/kit/types/index.d.ts (hunk 1 above)
Comment (Rich-Harris) [reply packages/kit/types/index.d.ts:639]: Is there any real drawback? It's contrived but you could imagine someone doing this sort of thing, where `some_shared_logic` is picky about the second argument:


```js
export const actions = {
  blah: ({ request, route }) => {
    some_shared_logic(request, route.id);
  }
};
```

**Comment by Conduitry** [comment]
Does that mean this also closes #6167?

**Comment by dummdidumm** [comment]
no, because that relates to `$page.routeId`, but you are probably right that this can also be infered from that.

**Comment by Rich-Harris** [comment]
oops, didn't add the tracking part yet. ignore that i marked this PR ready

**Comment by Conduitry** [comment]
FYI you can mark PRs as draft again. It's part of the "Reviewers" section in the upper right ("Still in progress? Convert to draft") which yes is an insane place for this to be.

**Comment by Rich-Harris** [comment]
Interesting wrinkle I found while adding a test: this...

```js
/** @type {import('./$types').LayoutServerLoad} */
export function load({ route }) {
  return { route };
}
```

...doesn't count as 'using' `route.id` even if the page renders it:

```svelte
<script>
  /** @type {import('./$types').LayoutData} */
  export let data;
</script>

<h1>route.id: {data.route.id}</h1>
<slot />
```

That's because the `uses` object is updated too late (after rendering/serialization). Need to thunkify it or something

**Comment by dummdidumm** [comment]
Doesn't the same apply to params and url as well? Not sure if we need to optimize for this.

**Comment by Rich-Harris** [comment]
Ok that was a bit of a faff but it will now correctly identify the dependency on `route.id` if you have this...

```js
/** @type {import('./$types').LayoutServerLoad} */
export function load({ route }) {
  return { route };
}
```

.,..and the same goes for `url` and `params`. Surprised we didn't get bitten by that bug sooner

**Review by Rich-Harris — APPROVED** [review]

**Pull request merged**